File: clinicflow-backend/src/models/Appointment.ts
import mongoose, { Schema, Document } from 'mongoose';
import aggregatePaginate from 'mongoose-aggregate-paginate-v2';

export interface IAppointment extends Document {
    patientId?: mongoose.Types.ObjectId;
    patientName?: string;
    patientEmail?: string;
    patientPhone?: string;
    doctorId: mongoose.Types.ObjectId;
    hospitalId: mongoose.Types.ObjectId;
    appointmentDate: Date;
    startTime: string;
    endTime: string;
    status: 'scheduled' | 'confirmed' | 'in-progress' | 'completed' | 'cancelled' | 'no-show';
    type: 'consultation' | 'follow-up' | 'emergency' | 'checkup';
    reason: string;
    notes: string;
    fee: number;
    paymentStatus: 'pending' | 'paid' | 'refunded';
    tokenNumber: string;
    createdBy?: mongoose.Types.ObjectId;
    createdAt: Date;
    updatedAt: Date;
}

const AppointmentSchema = new Schema<IAppointment>(
    {
        patientId: {
            type: Schema.Types.ObjectId,
            ref: 'Patient',
            required: false,
        },
        patientName: {
            type: String,
            required: false, // Required if patientId is missing, handled in validation
        },
        patientEmail: {
            type: String,
            required: false,
        },
        patientPhone: {
            type: String,
            required: false,
        },
        doctorId: {
            type: Schema.Types.ObjectId,
            ref: 'Doctor',
            required: true,
        },
        hospitalId: {
            type: Schema.Types.ObjectId,
            ref: 'Hospital',
            required: true,
        },
        appointmentDate: {
            type: Date,
            required: true,
        },
        startTime: {
            type: String,
            required: true,
        },
        endTime: {
            type: String,
            required: true,
        },
        status: {
            type: String,
            enum: ['scheduled', 'confirmed', 'in-progress', 'completed', 'cancelled', 'no-show'],
            default: 'scheduled',
        },
        type: {
            type: String,
            enum: ['consultation', 'follow-up', 'emergency', 'checkup'],
            default: 'consultation',
        },
        reason: {
            type: String,
            required: true,
        },
        notes: {
            type: String,
            default: '',
        },
        fee: {
            type: Number,
            default: 0,
        },
        paymentStatus: {
            type: String,
            enum: ['pending', 'paid', 'refunded'],
            default: 'pending',
        },
        tokenNumber: {
            type: String,
            required: false, // Will be auto-generated
        },
        createdBy: {
            type: Schema.Types.ObjectId,
            ref: 'User',
            required: false,
        },
    },
    {
        timestamps: true,
    }
);

AppointmentSchema.plugin(aggregatePaginate);

AppointmentSchema.index({ hospitalId: 1, appointmentDate: 1 });
AppointmentSchema.index({ patientId: 1 });
AppointmentSchema.index({ doctorId: 1, appointmentDate: 1 });
AppointmentSchema.index({ status: 1 });

export const Appointment = mongoose.model<IAppointment>('Appointment', AppointmentSchema);

File: server-rust/src/models/appointment.rs
use serde::{Deserialize, Serialize};
use bson::{oid::ObjectId, DateTime as BsonDateTime};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Appointment {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    #[serde(rename = "patientId", skip_serializing_if = "Option::is_none")]
    pub patient_id: Option<ObjectId>,
    #[serde(rename = "patientName", skip_serializing_if = "Option::is_none")]
    pub patient_name: Option<String>,
    #[serde(rename = "patientEmail", skip_serializing_if = "Option::is_none")]
    pub patient_email: Option<String>,
    #[serde(rename = "patientPhone", skip_serializing_if = "Option::is_none")]
    pub patient_phone: Option<String>,
    #[serde(rename = "doctorId")]
    pub doctor_id: ObjectId,
    #[serde(rename = "hospitalId")]
    pub hospital_id: ObjectId,
    #[serde(rename = "appointmentDate")]
    pub appointment_date: BsonDateTime,
    #[serde(rename = "startTime")]
    pub start_time: String,
    #[serde(rename = "endTime")]
    pub end_time: String,
    #[serde(default = "default_status")]
    pub status: String,
    #[serde(rename = "type", default = "default_type")]
    pub appointment_type: String,
    #[serde(default)]
    pub reason: String,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub fee: f64,
    #[serde(rename = "paymentStatus", default = "default_payment_status")]
    pub payment_status: String,
    #[serde(rename = "tokenNumber", skip_serializing_if = "Option::is_none")]
    pub token_number: Option<String>,
    #[serde(rename = "createdAt", skip_serializing_if = "Option::is_none")]
    pub created_at: Option<BsonDateTime>,
    #[serde(rename = "updatedAt", skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<BsonDateTime>,
}

fn default_status() -> String { "scheduled".to_string() }
fn default_type() -> String { "consultation".to_string() }
fn default_payment_status() -> String { "pending".to_string() }
